lib/sshutils/sftp/sftp_test.go

```diff
@@ -0,0 +1,435 @@
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package sftp
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/gravitational/teleport/lib/utils"
+
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	fileMaxSize = 1000
+)
+
+func TestMain(m *testing.M) {
+	utils.InitLoggerForTests()
+	os.Exit(m.Run())
+}
+
+func TestUpload(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name        string
+		srcPaths    []string
+		dstPath     string
+		opts        Options
+		files       []string
+		expectedErr string
+	}{
+		{
+			name: "one file",
+			srcPaths: []string{
+				"file",
+			},
+			dstPath: "copied-file",
+			opts: Options{
+				PreserveAttrs: true,
+			},
+			files: []string{
+				"file",
+			},
+		},
+		{
+			name: "one file to dir",
+			srcPaths: []string{
+				"file",
+			},
+			dstPath: "dst/",
+			opts: Options{
+				PreserveAttrs: true,
+			},
+			files: []string{
+				"file",
+				"dst/",
+			},
+		},
+		{
+			name: "one dir",
+			srcPaths: []string{
+				"src/",
+			},
+			dstPath: "dir/",
+			opts: Options{
+				PreserveAttrs: true,
+				Recursive:     true,
+			},
+			files: []string{
+				"src/",
+			},
+		},
+		{
+			name: "two files dst doesn't exist",
+			srcPaths: []string{
+				"src/file1",
+				"src/file2",
+			},
+			dstPath: "dst/",
+			opts: Options{
+				PreserveAttrs: true,
+			},
+			files: []string{
+				"src/file1",
+				"src/file2",
+			},
+		},
+		{
+			name: "two files dst does exist",
+			srcPaths: []string{
+				"src/file1",
+				"src/file2",
+			},
+			dstPath: "dst/",
+			opts: Options{
+				PreserveAttrs: true,
+			},
+			files: []string{
+				"src/file1",
+				"src/file2",
+				"dst/",
+			},
+		},
+		{
+			name: "nested dirs",
+			srcPaths: []string{
+				"s",
+			},
+			dstPath: "dst/",
+			opts: Options{
+				PreserveAttrs: true,
+				Recursive:     true,
+			},
+			files: []string{
+				"s/",
+				"s/file",
+				"s/r/",
+				"s/r/file",
+				"s/r/c/",
+				"s/r/c/file",
+				"dst/",
+			},
+		},
+		{
+			name: "multiple src dst not dir",
+			srcPaths: []string{
+				"uno",
+				"dos",
+				"tres",
+			},
+			dstPath: "dst_file",
+			files: []string{
+				"uno",
+				"dos",
+				"tres",
+				"dst_file",
+			},
+			expectedErr: `local file "%s/dst_file" is not a directory, but multiple source files were specified`,
+		},
+		{
+			name: "src dir with recursive not passed",
+			srcPaths: []string{
+				"src/",
+			},
+			dstPath: "dst/",
+			files: []string{
+				"src/",
+			},
+			expectedErr: `"%s/src" is a directory, but the recursive option was not passed`,
+		},
```

Comment:
As we talked earlier, about `~` support. Can you also add a test here to verify the home directory support?